### ISSUE TITLE
Installer tries to read non-existent directory when run from a symlink dir

### ISSUE THREAD

**Issue by stlaz** [openshift/installer#1839]
# Version

```console
$ openshift-install version
./bin/openshift-install unreleased-master-1110-g7cc42fa81b7a26c6ae180af3ff097f32d8c30c51
built from commit 7cc42fa81b7a26c6ae180af3ff097f32d8c30c51
release image registry.svc.ci.openshift.org/origin/release:4.2

```
(built from master)

# Platform (aws|libvirt|openstack):

aws

# What happened?

```console
$ ./bin/openshift-install create cluster --dir standovo
? SSH Public Key /home/slaznick/.ssh/aws_clusters.pub
? Platform aws
? Region us-east-1
? Base Domain devcluster.openshift.com
? Cluster Name slaznick-somename
? Pull Secret [? for help] <here be pull secret>

INFO Creating infrastructure resources...         
ERROR There are some problems with the configuration, described below. 
ERROR                                              
ERROR The Terraform configuration must be valid before initialization so that 
ERROR Terraform can determine which modules and providers need to be installed. 
ERROR                                              
ERROR Error: Failed to read module directory       
ERROR                                              
ERROR Module directory ../../../../../tmp/openshift-install-103911214 does not exist 
ERROR or cannot be read.                           
ERROR                                              
FATAL failed to fetch Cluster: failed to generate asset "Cluster": failed to create cluster: failed to initialize Terraform
```

# What you expected to happen?

Installation success.

# How to reproduce it (as minimally and precisely as possible)?
reproducible always:
1. `ln -s $HOME/go/src/github.com/openshift/ $HOME/projects/`
2. `cd $HOME/projects/openshift/installer`
3. `./hack/build.sh`
4. `./bin/openshift-install create cluster --dir something`
5. supply whatever info needed

**Comment by stlaz** [comment]
## Additional information
This happens when the installer is run from a symlink dir, I'm pretty sure such a use-case worked before. Trying to find a relative path (`../../../../../tmp/openshift-install-103911214` in the example) to a directory created _somwhere_ on a fs is quite unconventional and does not seem to be working well.

**Comment by Paul1975666** [comment]
I don't know anything about any of these questions or comments

**Comment by johwes** [comment]
I got this problem today as well and it has not behaved like this before. My homedir is under  "/home/remote/username" with a symlink "/home/username".

**Comment by wking** [comment]
The directory is [generated with `ioutil.TempDir`][1].  See Go docs [here][2] and [here][3] about the backing choices.  What is yoir `$TMPDIR`?

[1]: https://github.com/openshift/installer/blob/d9a9648cf2330d467cca9f2988846d031464125e/pkg/asset/cluster/cluster.go#L58
[2]: https://golang.org/pkg/io/ioutil/#TempDir
[3]: https://golang.org/pkg/os/#TempDir

**Comment by stlaz** [comment]
`$TMPDIR` is the default, thus unset. There's no problem with how the tempdir is created.

The problem is that the installer is trying to reach the tmp dir using a _relative path_ for reasons unknown from the _real path_ of where it's run from, which won't work if your symlinked directory has different level of nesting within the fs than the path it is pointing to.

Since you pointed us to the code dealing with this, the problem, as identified by the error string, gotta be somewhere in the lib call [here](https://github.com/openshift/installer/blob/d9a9648cf2330d467cca9f2988846d031464125e/pkg/asset/cluster/cluster.go#L73)

**Comment by lukaszachy** [comment]
I have the same issue on Linux (Fedora 30 , v4.2.4 built from commit 425e4ff0037487e32571258640b39f56d5ee5572)

'real' location is /home/lzachar/_important_/STH/bin/openshift-installer 
but I was using it as  /home/lzachar/STH/bin/openshift-installer (so it is ~/_important_/STH -> ~/STH link) 

And unless I run it inside bin dir (pwd = /home/lzachar/STH/bin/) I end up with ../../../../../ error - Number of ../ was one level off - from symlinked dir, but matching 'real' dir. All ../ went up to the root (/) dir.

--dir was located in symlinked structure as well

**Comment by patrickdillon** [comment]
It would be good to know the version of go used to build the installer. In discussion with @sanchezl he mentioned golang 1.12 has a bug evaluating relative symlinks. it should be fixed in go 1.12.1. References: 

https://go-review.googlesource.com/c/go/+/165197/

https://github.com/golang/go/issues?q=milestone%3AGo1.12.1

**Comment by patrickdillon** [comment]
This is an upstream bug in terraform: https://github.com/hashicorp/terraform/issues/24196

PR: https://github.com/hashicorp/terraform/pull/24197

We will look into carrying the patch locally
